### DISCUSSION TITLE
Typecast var to double

### DISCUSSION THREAD

**Discussion by Manushanker** [autodiff/autodiff#292] [General]
I am trying use reverse mode differentiator to compute higher order derivatives.  I use var data type and in the function I use a int cast operator to convert data into integer type. Please see the below code
static var  a0;
static int n0;
n0 = (int) (a0 * 1.1f) + 1;

The error message from the compiler is "E0413	no suitable conversion function from "autodiff::detail::ExprPtr<double>" to "int" exists"
May I know if there is a function in Autodiff to convert data type to double?

**Comment by allanleal** [comment]
Hi! `static_cast<double>(a0)` should work. 

For forward types (`dual` and `real`), you could also define `AUTODIFF_ENABLE_IMPLICIT_CONVERSION_DUAL` or `AUTODIFF_ENABLE_IMPLICIT_CONVERSION_REAL`. At the moment, this is not done for `var`. A similar preprocessor check would have to be done around the lines of code:

~~~cpp
/// Explicitly convert this Variable object into its underlying arithmetic type.
explicit operator T() const { return expr->val; }
~~~

in `autodiff/reverse/var/var.hpp`. This is what exists for `dual`, for example:

~~~cpp
    /// Convert this Dual number into a value of type @p U.
#if defined(AUTODIFF_ENABLE_IMPLICIT_CONVERSION_DUAL) || defined(AUTODIFF_ENABLE_IMPLICIT_CONVERSION)
    operator T() const { return val; }

    template<typename U>
    operator U() const { return static_cast<U>(val); }
#else
    explicit operator T() const { return val; }

    template<typename U>
    explicit operator U() const { return static_cast<U>(val); }
#endif
~~~

**Reply by Manushanker** [reply]
Thanks Allan it works! The static_cast<double>(a0) helped to resolve this issue. 
I have a struct like the one below
struct PEM {
    int pid;
    int mid;
    int mid2;                           // mid for surface impedance
    int ibc;                            // impedance boundary condition flag
    int cfie;                           // =1: CFIE, =0: EFIE
    var thick;
    var area;                        // <0.5: Skin effect not considered, >0.5 <1.5: Skin effect (double skin depth) considered for surface impedance
    COMPLEX2<var> zs;                        // surface impedance of impedance boundary condition
};
I use p_pem = MALLOC(PEM, n_pem + 1); to allocate memory for the variable p_pem. Malloc is allocating memory but for some reason it is not allocating for variables thick and area which are of type var. The same code works without issue when using dual in forward mode. 
Can you please provide a solution to fix it?
